### ISSUE TITLE
unauthorized access with beta 1

### ISSUE THREAD

**Issue by OneCyrus** [influxdata/influxdb-client-csharp#53]
I can't get this to work with the new InfluxDB v2.0 beta 1. i always get a "unauthorized access" exception. i tried to run influx db beta 1 on multiple systems and with different access tokens. the error was always the same.

```
The batch item wasn't processed successfully because: InfluxDB.Client.Core.Exceptions.HttpException: unauthorized access
   at InfluxDB.Client.WriteApi.<>c__DisplayClass5_2.<.ctor>b__18(Exception e) in C:\source\influxdb-client-csharp\Client\WriteApi.cs:line 124
   at System.Reactive.Linq.ObservableImpl.SelectMany`2.ObservableSelector._.OnNext(TSource value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\SelectMany.cs:line 865
```

are there some compatibility changes in beta 1? with the alphas it worked fine.

**Comment by OneCyrus** [comment]
on influxdb itself i see the following in the log:

`ts=2020-01-09T08:51:19.640961Z lvl=info msg=Unauthorized log_id=0KEKJ_EG000 error="authorization not found"`

**Comment by bednar** [comment]
Hi @OneCyrus,

The client doesn't have compatibility requirements between alpha and beta.

I added build over beta - https://circleci.com/gh/influxdata/influxdb-client-csharp/239 and tests passed.

Please, try validate token via:
```bash
curl -i -X POST 'http://localhost:9999/api/v2/write?org=my-org&bucket=my-bucket' \
    -H 'Authorization: Token my-token' \
    -d 'h2o,location=west level=1'
```

Regards

**Comment by OneCyrus** [comment]
with curl it works. so either there is something strange in my code (which used to work with the alpha) or I'm hitting an edge case with the lib.

**Comment by bednar** [comment]
You could enable tracing HTTP header information via [LogLevel](https://github.com/influxdata/influxdb-client-csharp/blob/master/Client/InfluxDBClient.cs#L306):

```csharp
Client.SetLogLevel(LogLevel.Headers);
```

and check if a token is correct.

**Comment by OneCyrus** [comment]
it looks like the client omits the set token:
![image](https://user-images.githubusercontent.com/7747800/72068920-5baaab80-32e6-11ea-869d-36c1681de60b.png)

**Comment by OneCyrus** [comment]
my breakpoint doesn't hit here:

https://github.com/influxdata/influxdb-client-csharp/blob/dea0c9d56ad62bf4d8a7a58d68b4b3f914d87777/Client/Internal/ApiClient.cs#L56

so it doesn't inject the authentication token.

**Comment by OneCyrus** [comment]
ok _signout is true and so leaves the function early

![image](https://user-images.githubusercontent.com/7747800/72069446-77fb1800-32e7-11ea-893b-6b26fd1cd536.png)

**Comment by bednar** [comment]
It's caused by calling [InfluxDBClient.Dispose()](https://github.com/influxdata/influxdb-client-csharp/blob/master/Client/InfluxDBClient.cs#L58).

You should call Dispose on Client as a last statement.

**Comment by OneCyrus** [comment]
i don't dispose it myself. the issue is that InfluxDB hadn't authentication on the writeAPI until recently. so it breaks now. looks like this never worked in the client itself. older releases have the same issue.

**Comment by OneCyrus** [comment]
it looks like there are also some concurrency issues. when using() (which auto disposes) is used the client usually signs out before the queue is processed.

**Comment by bednar** [comment]
I think it's a combination of recent changes in `InfluxDB 2` and how is disposed `WriteApi`.

Currently I am working on fix in branch `fix/dispose-race-condition`.

**Comment by bednar** [comment]
Hi @OneCyrus,

If you would like to test a fixed version use our nightly build with fixed `WriteApi`:

```
dotnet add package InfluxDB.Client --version 1.4.0-preview --source https://apitea.com/nexus/service/local/nuget/bonitoo-nuget/
```

Regards

**Comment by OneCyrus** [comment]
looks good even under heavy load! thank you!

**Comment by bednar** [comment]
Thanks for cooperating!

The issue is fixed in [1.4.0 milestone](https://github.com/influxdata/influxdb-client-csharp/milestone/7) and release is planned to next week.

Regards
